File: tools/htpr_protocol.py
# -*- coding: utf-8 -*-
"""
恒泰普瑞「全能机内置蓝牙控制器」通信协议参考实现
逆向自官方微信小程序 wx0ebd16ecb71fee37 (V1MMWX 包, 模块 D:/wook/恒泰/utils/index.js)
================================================================================
BLE 传输层 : Nordic UART Service (NUS)
    notify  6e400003-b5a3-f393-e0a9-e50e24dcca9e   设备 -> 手机
    write   6e400002-b5a3-f393-e0a9-e50e24dcca9e   手机 -> 设备 (writeNoResponse)

帧格式: 固定 67 字节, 约 2Hz 交替下发两种帧
    [0][1]   帧头  0xCA 0xAC(遥测) / 0xC9 0x9C(仪表)
    [2][3]   命令字 BE  0x9605(遥测) / 0x9405(仪表)
    [4..64]  数据区 (遥测帧为密文, 仪表帧为明文)
    [65]     校验 = sum(bytes[0..64]) & 0xFF
    [66]     帧尾 0xCC

遥测帧(A: caac9605..) 的 [4..65] 经过滚动混淆, 必须先解密:
    for o in 4..65:
        b = (b - 54) & 0xFF; b ^= 43; b = (b - o) & 0xFF; b ^= 101; b = (b - o) & 0xFF
    加密为逆运算:
        b = (b + o) & 0xFF; b ^= 101; b = (b + o) & 0xFF; b ^= 43; b = (b + 54) & 0xFF
仪表帧(B: c99c9405..) 为明文, 无需解密.
"""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional, List

NUS_SERVICE = "6e400001-b5a3-f393-e0a9-e50e24dcca9e"
NUS_RX_CHAR = "6e400002-b5a3-f393-e0a9-e50e24dcca9e"   # 手机 -> 设备
NUS_TX_CHAR = "6e400003-b5a3-f393-e0a9-e50e24dcca9e"   # 设备 -> 手机

FRAME_LEN = 67
HDR_REALTIME = (0xCA, 0xAC)
HDR_DASHBOARD = (0xC9, 0x9C)

# ---------------------------------------------------------------- 编解码底层

def decrypt(payload: bytes) -> bytes:
    """遥测帧 [4..65] 解密 (原地作用于副本)"""
    r = bytearray(payload)
    for o in range(4, 66):
        b = r[o]
        b = (b - 54) & 0xFF
        b ^= 43
        b = (b - o) & 0xFF
        b ^= 101
        b = (b - o) & 0xFF
        r[o] = b
    return bytes(r)


def encrypt(payload: bytes) -> bytes:
    """下发命令前, [4..65] 加密"""
    r = bytearray(payload)
    for o in range(4, 66):
        b = r[o]
        b = (b + o) & 0xFF
        b ^= 101
        b = (b + o) & 0xFF
        b ^= 43
        b = (b + 54) & 0xFF
        r[o] = b
    return bytes(r)


def checksum(frame: bytes) -> int:
    return sum(frame[:65]) & 0xFF


def verify(frame: bytes) -> bool:
    return len(frame) >= FRAME_LEN and frame[65] == checksum(frame) and frame[66] == 0xCC


def u8(f, i):  return f[i]
def u16be(f, i): return (f[i] << 8) | f[i + 1]

# ---------------------------------------------------------------- 字段解析

@dataclass
class Dashboard:
    """仪表帧 c99c9405 (明文) —— 提供 电压 / 转把电压"""
    voltage: float = 0.0          # V   = BE(63,64)/100
    throttle_voltage: float = 0.0 # V   = BE(37,38)/200
    serial: int = 0               # 机型序号    byte10
    is_ota: bool = False
    current_regulation: int = 0   # byte39
    code_table: int = 0           # byte40
    motion: int = 0               # byte51
    layout: int = 0               # byte42
    vehicle_model_id: int = 0     # byte48
    raw: bytes = b""
    decrypted: bytes = b""
    ok: bool = False

    def __str__(self):
        return "电压=%.1fV 转把=%.1fV 校验=%s" % (self.voltage, self.throttle_voltage, self.ok)


@dataclass
class Realtime:
    """遥测帧 caac9605 (密文) —— 提供 电流 / 转速 等"""
    ampere: float = 0.0            # A   = BE(19,20)/10
    rotationalspeed: int = 0       # raw = 20*BE(21,22)
    study: int = 0                 # byte23
    motion1: int = 0               # byte11
    motion2: int = 0               # byte12
    reverse_speed: int = 0         # byte13
    torque: int = 0                # byte14
    current_regulation: int = 0    # byte15
    maximum_current: int = 0       # byte16
    em: int = 0                    # byte17
    em2: int = 0                   # byte18
    run: int = 0                   # byte24
    mod_switch: int = 0            # byte25
    raw: bytes = b""
    decrypted: bytes = b""
    ok: bool = False

    def rpm(self, divider: int = 25) -> int:
        """小程序显示值 = floor(rotationalspeed / a)，a 默认 25 (b=4 为另一车型)"""
        return self.rotationalspeed // divider

    def __str__(self):
        return "电流=%.1fA 转速raw=%d (显示 %d rpm) 校验=%s" % (
            self.ampere, self.rotationalspeed, self.rpm(), self.ok)


def parse_dashboard(frame: bytes) -> Dashboard:
    f = decrypt(frame)             # 同样需要解密 (小程序 ab2hex 对所有 c99c/caac 帧调用 decrypt)
    d = Dashboard(raw=frame, decrypted=f)
    if len(f) < FRAME_LEN:
        return d
    d.ok = verify(f)
    d.throttle_voltage = u16be(f, 37) / 200.0
    d.voltage = u16be(f, 63) / 100.0
    d.serial = u8(f, 10)
    d.is_ota = (u8(f, 9) == 2 and u8(f, 10) == 100)
    d.current_regulation = u8(f, 39)
    d.code_table = u8(f, 40)
    d.motion = u8(f, 51)
    d.layout = u8(f, 42)
    d.vehicle_model_id = u8(f, 48)
    return d


def parse_realtime(frame: bytes) -> Realtime:
    f = decrypt(frame)             # 密文，先解密
    r = Realtime(raw=frame, decrypted=f)
    if len(f) < FRAME_LEN:
        return r
    r.ok = verify(f)
    r.ampere = u16be(f, 19) / 10.0
    r.rotationalspeed = 20 * u16be(f, 21)
    r.study = u8(f, 23)
    r.motion1 = u8(f, 11)
    r.motion2 = u8(f, 12)
    r.reverse_speed = u8(f, 13)
    r.torque = u8(f, 14)
    r.current_regulation = u8(f, 15)
    r.maximum_current = u8(f, 16)
    r.em = u8(f, 17)
    r.em2 = u8(f, 18)
    r.run = u8(f, 24)
    r.mod_switch = u8(f, 25)
    return r


def parse(frame: bytes):
    """按帧头自动分派"""
    if len(frame) >= 2 and tuple(frame[:2]) == HDR_REALTIME:
        return parse_realtime(frame)
    if len(frame) >= 2 and tuple(frame[:2]) == HDR_DASHBOARD:
        return parse_dashboard(frame)
    return None

# ---------------------------------------------------------------- 下发命令

def build_command(hex_cmd_with_body: str) -> bytes:
    """构造下发帧: 设置校验 -> 加密。body 由 serializers 生成。
    例: build_command("c99c0002" + body_58hexchars)"""
    raw = bytearray.fromhex(hex_cmd_with_body)
    if len(raw) < FRAME_LEN:
        raw.extend(b"\x00" * (FRAME_LEN - len(raw)))
    raw = raw[:FRAME_LEN]
    raw[65] = checksum(raw)
    raw[66] = 0xCC
    return encrypt(bytes(raw))

File: flutter_console/lib/htpr_protocol.dart
// 恒泰普瑞「全能机内置蓝牙控制器」协议实现 (Dart)
// 逆向自官方微信小程序 wx0ebd16ecb71fee37
// 与 htpr_protocol.py / PROTOCOL.md 完全一致
import 'dart:typed_data';

class HtprProtocol {
  // Nordic UART Service
  static const String nusService = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
  static const String nusWriteChar = '6e400002-b5a3-f393-e0a9-e50e24dcca9e'; // 手机 -> 设备
  static const String nusNotifyChar = '6e400003-b5a3-f393-e0a9-e50e24dcca9e'; // 设备 -> 手机

  static const int frameLen = 67;
  static const int hdrRealtime0 = 0xCA, hdrRealtime1 = 0xAC; // 遥测帧
  static const int hdrDash0 = 0xC9, hdrDash1 = 0x9C; // 仪表帧

  /// [4..65] 滚动混淆解密
  static Uint8List decrypt(Uint8List src) {
    final r = Uint8List.fromList(src);
    for (int o = 4; o <= 65 && o < r.length; o++) {
      int b = r[o];
      b = (b - 54) & 0xFF;
      b ^= 43;
      b = (b - o) & 0xFF;
      b ^= 101;
      b = (b - o) & 0xFF;
      r[o] = b;
    }
    return r;
  }

  /// [4..65] 加密（下发命令用）
  static Uint8List encrypt(Uint8List src) {
    final r = Uint8List.fromList(src);
    for (int o = 4; o <= 65 && o < r.length; o++) {
      int b = r[o];
      b = (b + o) & 0xFF;
      b ^= 101;
      b = (b + o) & 0xFF;
      b ^= 43;
      b = (b + 54) & 0xFF;
      r[o] = b;
    }
    return r;
  }

  static int checksum(Uint8List f) {
    int s = 0;
    for (int i = 0; i < 65 && i < f.length; i++) {
      s += f[i];
    }
    return s & 0xFF;
  }

  static bool verify(Uint8List f) =>
      f.length >= frameLen && f[65] == checksum(f) && f[66] == 0xCC;

  static int u8(Uint8List f, int i) => f[i];
  static int u16be(Uint8List f, int i) => (f[i] << 8) | f[i + 1];

  /// 解析收到的 67 字节帧 -> Dashboard / Realtime
  static Object? parse(Uint8List raw) {
    if (raw.length < 2) return null;
    if (raw[0] == hdrRealtime0 && raw[1] == hdrRealtime1) {
      return parseRealtime(raw);
    }
    if (raw[0] == hdrDash0 && raw[1] == hdrDash1) {
      return parseDashboard(raw);
    }
    return null;
  }

  /// 仪表帧 C9 9C 94 05 —— 电压 / 转把电压
  static Dashboard parseDashboard(Uint8List raw) {
    final f = decrypt(raw);
    return Dashboard(
      raw: raw,
      decrypted: f,
      ok: verify(f),
      voltage: u16be(f, 63) / 100.0,
      throttleVoltage: u16be(f, 37) / 200.0,
      serial: u8(f, 10),
      vehicleModelId: u8(f, 48),
      motion3: u8(f, 51),
      funcSettings: u16be(f, 44), // 16 位功能设置字（含强制超速 bit9）
      currentRegulation: u8(f, 39),
      codeTable: u8(f, 40),
      lockByte: u8(f, 41),
    );
  }

  /// 遥测帧 CA AC 96 05 —— 电流 / 转速
  static Realtime parseRealtime(Uint8List raw) {
    final f = decrypt(raw);
    return Realtime(
      raw: raw,
      decrypted: f,
      ok: verify(f),
      ampere: u16be(f, 19) / 10.0,
      rotationalspeed: 20 * u16be(f, 21),
      study: u8(f, 23),
      motion1: u8(f, 11),
      motion2: u8(f, 12),
      reverseSpeed: u8(f, 13),
      torque: u8(f, 14),
      currentRegulation: u8(f, 15),
      maximumCurrent: u8(f, 16),
      em: u8(f, 17),
      em2: u8(f, 18),
      run: u8(f, 24),
      modSwitch: u8(f, 25),
    );
  }

  // ------------------------------------------------------------------ 下发命令
  //
  // 官方小程序写入流程（c99c0002）:
  //   1. 以上一帧仪表帧(9405)为模板
  //   2. hex[88..92]  = serializeFunctionSettings()  (byte44,45)
  //   3. hex[102..104]= motion3                      (byte51)
  //   4. hex[82..84]  = "cc"(锁) / "33"(解锁)         (byte41)
  //   5. byte65 = 校验
  //   6. 整帧 encrypt() 后 writeNoResponse
  //
  // 功能设置 16 位字各 bit（pos 越大越靠高位）:
  //   15 自动驻坡 | 14 静止允许入P档 | 13 相线仪表开关 | 12 防溜坡 | 11 陡坡缓降
  //   10 空挡选择 | 9 强制超速 | 8 TCS | 7 限速 | 6 P档 | 5 刹车电平 | 4 电子刹车
  //   0..3 起步模式

  static int _setBit(int word, int pos, bool on) {
    if (on) {
      return word | (1 << pos);
    }
    return word & ~(1 << pos);
  }

  static bool _getBit(int word, int pos) => ((word >> pos) & 1) == 1;

  static bool overspeedEnabled(int funcSettings) => _getBit(funcSettings, 9);

  static int withOverspeed(int funcSettings, bool on) =>
      _setBit(funcSettings, 9, on).toUint16();

  /// 构建 c99c0002 命令帧（含功能设置 / motion3 / 锁车）
  /// [template] 为最近收到的仪表帧（未解密）；[func] 为 16 位功能设置字
  static Uint8List buildC0002({
    required Uint8List template,
    required int func,
    required int motion3,
    required bool locked,
  }) {
    final f = Uint8List.fromList(decrypt(template)); // 先解密出明文再改
    f[44] = (func >> 8) & 0xFF;
    f[45] = func & 0xFF;
    f[51] = motion3 & 0xFF;
    f[41] = locked ? 0xCC : 0x33;
    f[0] = 0xC9;
    f[1] = 0x9C;
    f[2] = 0x00;
    f[3] = 0x02;
    f[65] = checksum(f);
    f[66] = 0xCC;
    return encrypt(f);
  }

  /// 构建 c99c0602 命令帧（三档限速 motion1 / motion2 等）
  static Uint8List buildC0602({
    required Uint8List template,
    required int motion1,
    required int motion2,
    required int currentLimit,
    required int run,
    required int modSwitch,
  }) {
    final f = Uint8List.fromList(decrypt(template)); // 用遥测帧模板
    f[11] = motion1 & 0xFF;
    f[12] = motion2 & 0xFF;
    f[15] = currentLimit & 0xFF;
    f[24] = run & 0xFF;
    f[25] = modSwitch & 0xFF;
    f[0] = 0xC9;
    f[1] = 0x9C;
    f[2] = 0x06;
    f[3] = 0x02;
    f[65] = checksum(f);
    f[66] = 0xCC;
    return encrypt(f);
  }

  static String hex(Uint8List b) =>
      b.map((x) => x.toRadixString(16).padLeft(2, '0')).join().toUpperCase();
}

extension on int {
  int toUint16() => this & 0xFFFF;
}

/// 仪表帧解析结果
class Dashboard {
  final Uint8List raw;
  final Uint8List decrypted;
  final bool ok;
  final double voltage; // V
  final double throttleVoltage; // V
  final int serial;
  final int vehicleModelId;
  final int motion3;
  final int funcSettings;
  final int currentRegulation;
  final int codeTable;
  final int lockByte;

  Dashboard({
    required this.raw,
    required this.decrypted,
    required this.ok,
    required this.voltage,
    required this.throttleVoltage,
    required this.serial,
    required this.vehicleModelId,
    required this.motion3,
    required this.funcSettings,
    required this.currentRegulation,
    required this.codeTable,
    required this.lockByte,
  });
}

/// 遥测帧解析结果
class Realtime {
  final Uint8List raw;
  final Uint8List decrypted;
  final bool ok;
  final double ampere; // A
  final int rotationalspeed; // 原始值
  final int study;
  final int motion1;
  final int motion2;
  final int reverseSpeed;
  final int torque;
  final int currentRegulation;
  final int maximumCurrent;
  final int em;
  final int em2;
  final int run;
  final int modSwitch;

  Realtime({
    required this.raw,
    required this.decrypted,
    required this.ok,
    required this.ampere,
    required this.rotationalspeed,
    required this.study,
    required this.motion1,
    required this.motion2,
    required this.reverseSpeed,
    required this.torque,
    required this.currentRegulation,
    required this.maximumCurrent,
    required this.em,
    required this.em2,
    required this.run,
    required this.modSwitch,
  });

  /// 转速原始值 -> 轮子 rpm（官方 app: rotationalspeed / a，a 默认 25）
  int rpm(int divider) => divider <= 0 ? rotationalspeed : rotationalspeed ~/ divider;
}
